### ISSUE TITLE
Some requirements for a new URI class

### ISSUE THREAD

**Issue by bblfish** [typelevel/cats-uri#6]
Having [tried lemonlabs Uri in banana-rdf](https://github.com/banana-rdf/banana-rdf/pull/379) and using http4s URi I have some ideas of key elements to improve http4s Uri. 

1. Http4s Uris are currently actually URLs, and should be named as such. Urls can have paths, be relativised, absolutized and be used for requests using something like an http framework. URNs cannot. We have `type URI = URL | URN` 
2. Http4s should take onboard the typing lemonlabs uses, that clearly distinguishes absolute Urls, relative Urls (and there relative path absolute: urls starting with `/`) and also schema relative urls.  That would allow Http4s to let us know if a Request is meant to take an absolute or a relative Url for example. There may be clients that can take both, but not all clients can. A client that has no connection cannot take a relative Url. A client bound to a connection can take a (path absolute) relative Url). 
Further crazy requirements:

1. Ideally (here we may be going beyond what is possible) it would be possible for the interface to be implemented so as to allow unique storage of URIs to re-use components. So one would be able to have a garbage collected store of URIs starting with "https" then followed by domain components, etc... so that Uris could be compare using `eq` for equality and use as little memory as possible. 
2. In RDF one has so many URIs the frameworks are careful to minimally parse them, leaving perhaps invalid uris in graphs rather than having to parse the structure of them. I wonder if there is a clever way of optionally integrating that requirement. 
3. Perhaps one also wants to map URIs to 64bit Long numbers. At the RDF layer one does not need to look into the URIs at all: one could just as well use 64bit long comparison to test for equality, once those are parsed and stored in a DB. 

Having an excellent real URI class that encompasses both URNs and URLs would be very helpful when writing application code. On the web application code proceeds as follows:
  1. client fetches hyper-text or hyper-data resource on the web (using http4s)
  2. it parses the content, often containing URIs (relative and absolute URLs and sometimes  even URNs) and may follow any number of those URLs in sequence or in parallel (going back to 1.). 
  
If the code could use the same URI structure in the parsing stage as it uses in the fetching stage that would make for much nicer and smoother coding. Note in RDF URIs are the core of the naming scheme, so one is using them all the time for every data structure. In banana-rdf I added support for relative urls, as those are useful when fetching data and also when posting graphs to a container: the server decides on the URL of the created resource and so the graph sent has to be one that can contain relative Urls. 

I also write this up on the http4s discussion area https://github.com/http4s/http4s/discussions/5930#discussioncomment-3777066

**Comment by isomarcte** [comment]
> Http4s Uris are currently actually URLs, and should be named as such. Urls can have paths, be relativised, absolutized and be used for requests using something like an http framework. URNs cannot. We have type URI = URL | URN

[
URLs aren't really a formal construct anymore, but more of an informal (and useful) way to discuss URIs which happen to refer to a network resource.](https://datatracker.ietf.org/doc/html/rfc3305#section-2.2) That said, cats-uri is planning to model URIs with scheme specific semantics for certain common schemes, such as http/https, so I think it will address the distinctions you are getting at. I plan to model URNs as well, but I'm not sure how far that will get before the initial release.

> Http4s should take onboard the typing lemonlabs uses, that clearly distinguishes absolute Urls, relative Urls (and there relative path absolute: urls starting with /) and also schema relative urls. That would allow Http4s to let us know if a Request is meant to take an absolute or a relative Url for example. There may be clients that can take both, but not all clients can. A client that has no connection cannot take a relative Url. A client bound to a connection can take a (path absolute) relative Url).

cats-uri will formally model relative and absolute URIs, so consumers will be able to determine that.

> Ideally (here we may be going beyond what is possible) it would be possible for the interface to be implemented so as to allow unique storage of URIs to re-use components. So one would be able to have a garbage collected store of URIs starting with "https" then followed by domain components, etc... so that Uris could be compare using eq for equality and use as little memory as possible.

This type of interning can be useful, but is not free. It requires some synchronization and computation. For this reason, I do not plan on supporting this in the library at this time. If needed, it can be added by consumers of this library.

> In RDF one has so many URIs the frameworks are careful to minimally parse them, leaving perhaps invalid uris in graphs rather than having to parse the structure of them. I wonder if there is a clever way of optionally integrating that requirement.

cats-uri will have formal modeling for all components of a URI, such as scheme/authority/path/query/etc. A consumer of the library will be able to use those to parse sub components of a String should they wish. I don't currently plan on a fuzzy parser for a whole URI.

> Perhaps one also wants to map URIs to 64bit Long numbers. At the RDF layer one does not need to look into the URIs at all: one could just as well use 64bit long comparison to test for equality, once those are parsed and stored in a DB.

Similar to the interning above, this is out of scope for the library itself. FWIW, this sounds like a hashing output and we can't universally uniquely map the full domain of URIs to a 64-bit range.

Thanks for your thoughts and suggestings @bblfish. I think you will be pretty happy with the direction we are going. cats-uri is going to be dependent on https://github.com/typelevel/case-insensitive 2.x.x and https://github.com/typelevel/idna4s 0.1.x, and I've been preoccupied in these dependent projects, but I hoping to have a RC out before too long. I want to get this proposed for http4s 1.x.x.

**Comment by bblfish** [comment]
Thanks for looking at the thoughts I put out there, most of them just thrown out in case it could help in designing the perfect Uri class :-) 

> > Http4s Uris are currently actually URLs, and should be named as such. Urls can have paths, be relativised, absolutized and be used for requests using something like an http framework. URNs cannot. We have type URI = URL | URN
> 
> [ URLs aren't really a formal construct anymore, but more of an informal (and useful) way to discuss URIs which happen to refer to a network resource.](https://datatracker.ietf.org/doc/html/rfc3305#section-2.2) 

As Type Safe/Level programmers and category theorists we know though that a Type is identified by the incoming and outgoing arrows. There is no concept of relative URN I think. One cannot resolve a relative URL with a URN, and one needs an Absolute URL to resolve a relative one. Similarly to relativize a URL to a base one needs an absolute URL. HTTP Requests take URLs (specific ones indeed: either http or https ones, but clearly it could be extended to some http3 url type if needed).
All that means we'll end up with approximately the types that lemonlabs ended up with. (Hopefully with a few less use of `null`s  or `Option`s for that matter :-) )

They do also say: 
>  An http URI is a  URL.  The phrase "URL scheme" is now used infrequently, usually to
   refer to some subclass of URI schemes which exclude URNs.

RDF **reasoning** really only needs the URI or IRI notion. Note that RDF reasoning does not occur in simple categories but in bicategories (see [Knowledge Representation in Bicategories of Relations](https://arxiv.org/abs/1706.00526), which are symmetric monoidal categories - the ones it turns out are useful for working on parallel processes and for working in quantum physics as eg. explained in [Picturing Quantum Processes](https://www.cambridge.org/core/books/picturing-quantum-processes/1119568B3101F3A685BE832FEEC53E52)). It turns out that Rel the category of Sets and Relations is equivalent to the Kleisli category of the Powerset Monad. Anyway looked at declaratively one only needs URIs.

But Pragmatically we need URLs and relative URLs because fetching a document is equivalent to a speech act - I call it a document act. And that means that we are positioned in the world and are working in a context (eg. that we are connected to a particular server). 

So I'd be a bit careful about what ended up being written in the RFC abstracts, because the RFC was a product very likely of these different points of views and interests coming together: the declarative and the pragmatic one. 

> That said, cats-uri is planning to model URIs with scheme specific semantics for certain common schemes, such as http/https, so I think it will address the distinctions you are getting at. I plan to model URNs as well, but I'm not sure how far that will get before the initial release.

great!

**Comment by bblfish** [comment]
I wrote:
> Hopefully with a few less use of nulls or Options for that matter :-) 

One way of working out what types (or subtypes in OO) would be useful is to see how far one can get without using any `null` or `Option`s. So for example it makes sense to have a `FragmentURL` type, that is a URL with a hash, and so have a type `FragmentlessURL` too perhaps. Where would that be useful? Well, for example, Http Requests should not be built with FragmentURLs - only with Fragmentless URLs. Note, there is no Fragment for URNs as per [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) diagram. 

```rfc
         foo://example.com:8042/over/there?name=ferret#nose
         \_/   \______________/\_________/ \_________/ \__/
          |           |            |            |        |
       scheme     authority       path        query   fragment
          |   _____________________|__
         / \ /                        \
         urn:example:animal:ferret:nose
```         

There may be other places in a header, eg. Link Header where both are acceptable, and indeed relative URLs are and even URNs. That can then be indicated by allowing generic Uris. Note one also would like the concept of IRIs to be around: looking at most APIs currently I can never be sure if toString is returning a URI string or IRI string...

**Comment by bblfish** [comment]
Btw. this thread could be turned into a discussion, if you open a discussion space.

Thinking a bit more about not using nulls I came across the following equivalences, that 
I think indicate that one should be able to improve over the wide use of nulls or options (eg. [lemonlabs.Uri](https://github.com/lemonlabsuk/scala-uri/blob/782917bd95b544f4cfec2fd99b2c973004d9b767/shared/src/main/scala/io/lemonlabs/uri/Uri.scala#L564) or [http4s.Uri](https://github.com/http4s/http4s/blob/series/0.23/core/shared/src/main/scala/org/http4s/Uri.scala#L67). Both present a URI as something like the following structure.

```scala
type URI = Option[Scheme] × Option[Authority] × Option[Path] × Option[Query] × Option[Fragment] 
```

Or in more mathematical formalism 

$URI =  (1 + S) × (1 + A) × (1 + P) × (1 + Q) × (1 + F)$

which we can expand slowly as follows:

$URI =  (1 + A + S + S×A) × (1 + P) × (1 + Q) × (1 + F)$
$URI = (1 + P +A + A×P + S + S×P + S×A + S×A×P) × (1 + Q) × (1 + F)$

We'll drop the × sign from here to make it fit

$URI = (1 + Q + P + PQ + A + AQ + AP + APQ + S + SQ + SP + SPQ + SA + SAQ + SAP + SAPQ) × (1 + F)$


$$
URI = 1 + F + Q + QF + P + PF + PQ + PQF + A + AF + AQ + AQF + AP + APF + APQ + APQF + S + SF + SQ + SQF + SP + SPF + SPQ + SPQF + SA + SAF + SAQ +  SAQF + SAP + SAPF + SAPQ + SAPQF
$$


So that's a hell of a disjunction! One should wonder immediately if all of these really make sense and how one could cut them down by pruning those that don't. Looking at it I don't think $A×F$ - an authority and a fragment - make sense. Neither does a $S×F$  a Scheme and a fragment.
Fragments are tied to Paths and Queries, not to schemes or authorities...

**Comment by bblfish** [comment]
One interesting thing that the arithmetic exercise above shows is that after one has pruned away the impossible combinations, I think one will find that one can define the relative URL without creating any new classes, just by type definition. 

So say we need the following class 
 + AgentChannel (eg. "https://bblfish.net:443") 
 + Path with subclass AbsolutePath (ie a Path that starts with /) which for simplicity would including queries (eg. `people/henry/card`)
 + Fragment
 
 Then we have a type (probably defined as a class or case class)
 ```
case class URL(ac: AgentChannel, path: AbsolutePath, frag: Option[Fragment])
 ```

but relative URLs need introduce nothing new. They can just be defined in terms of the components we already have

```
type RelativeURL = Path × (1 + Fragment) ≅ Path + (Path × Fragment)
```


So then URLs are always absolute URLs, and we can just use Paths when the AgentChannel is missing. That should work well with scala3's new disjunction `|` syntax, since URIs and Paths are already disjoint. If a function accepts relative URLs it can just take as argument `URL|Path`

Some examples:

If Http4s were to create a Request taking a relative URL, then it could have a constructor on `AbsolutePath` only, and would never want the Fragment, as that should not be passed over HTTP. 

Link relations in headers would take an argument `URL | RelativeURL` as defined above, since a link can be both of those (and perhaps even a URN - I would need to check). 

The header `Origin` would take only the `Host` part of what I called above the `AgentChannel` (I was thinking this may be what pi calculus calls a channel)




 
 Somehow this just feels very neat.
